### PR TITLE
Backport master to 4.4

### PR DESCRIPTION
**Pull request by cehbrecht**
# Overview

This branch backports the changes on master to pywps 4.4.

# Related Issue / Discussion

#568

# Additional Information

List of integrated PRs:

* Added validation for GPX files (#537)
* fix geojson mimetype (#529)
* bumpverion (#527)
* allow custom UOM reference for literal IO (#523)
* removed unused wpsserver.py (#521)
* skip py27 (#520)
* fix tests with language option (#519)
* Add NcML to supported FORMATS (#515)

Skipped PRs:
* jobqueue (#505)
* update docs for processes parameter (#518)

# Contribution Agreement

(as per https://github.com/geopython/pywps/blob/master/CONTRIBUTING.rst#contributions-and-licensing)

- [ ] I'd like to contribute [feature X|bugfix Y|docs|something else] to PyWPS. I confirm that my contributions to PyWPS will be compatible with the PyWPS license guidelines at the time of contribution.
- [ x] I have already previously agreed to the PyWPS Contributions and Licensing Guidelines

### CODE REVIEW

**Comment by cehbrecht** [comment]
test with emu ... still seams to work.

**Comment by cehbrecht** [comment]
@tlvu @Zeitsperre Maybe you like to check this PR? We can still improve on pywps-4.4 also after merge ...

**Comment by Zeitsperre** [comment]
@cehbrecht Would you like me to merge #573 into `pywps-4.4` before this is merged?

**Review by Zeitsperre — APPROVED** [review]
Looks good to me. For the Mmmetype list, I can't help but notice that the GeoTIFF mimetype is a matter of intense debate (https://github.com/opengeospatial/geotiff/issues/34). Does it make sense to include `Format('image/tiff', extension='.tiff', encoding='base64')` as well? What are your thoughts?

Also, cosmetic opinion, but would you be open to `isort`-ing the library imports to clean them up a bit?

File: pywps/inout/formats/__init__.py [hunk 1]
```diff
@@ -163,7 +163,7 @@ def json(self, jsonin):
 
 
 FORMATS = _FORMATS(
-    Format('application/vnd.geo+json', extension='.geojson'),
+    Format('application/geo+json', extension='.geojson'),
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply pywps/inout/formats/__init__.py:166]: I agree with this. Older code could still use the vdn variant and there is no harm in keeping it.

File: pywps/app/Process.py [hunk 2]
```diff
@@ -201,7 +200,7 @@ def _execute_process(self, async_, wps_request, wps_response):
             LOGGER.debug("Stored processes: {}".format(stored))
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply pywps/app/Process.py:200]: Also agree on that. Avoids the format delay (although marginal).

File: pywps/exceptions.py [hunk 3]
```diff
@@ -66,14 +65,14 @@ def get_response(self, environ=None):
...
-            u'<!-- PyWPS {version} -->\n'
-            u'<ows:ExceptionReport xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/ows/1.1 http://schemas.opengis.net/ows/1.1.0/owsExceptionReport.xsd" version="1.0.0">\n'  # noqa
-            u'  <ows:Exception exceptionCode="{name}" locator="{locator}" >\n'
-            u'      {description}\n'
-            u'  </ows:Exception>\n'
-            u'</ows:ExceptionReport>'
+        doc = str((
+            '<?xml version="1.0" encoding="UTF-8"?>\n'
+            '<!-- PyWPS {version} -->\n'
+            '<ows:ExceptionReport xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/ows/1.1 http://schemas.opengis.net/ows/1.1.0/owsExceptionReport.xsd" version="1.0.0">\n'  # noqa
```
Comment (fmigneault) [inline pywps/exceptions.py:71]: Shouldn't `version="1.0.0"` either be the actual WPS version, or correspond to `1.1.0` OWS schema?

File: setup.cfg [hunk 4]
```diff
@@ -1,4 +1,20 @@
 [flake8]
 ignore=F401,E402,W606
 max-line-length=120
-exclude=tests
+exclude=tests, pywps/alembic/versions
+
+[bumpversion]
+current_version = 4.2.11
```
Comment (fmigneault) [inline setup.cfg:7]: 4.4.x ?

**Comment by cehbrecht** [comment]
@Zeitsperre @fmigneault Thanks for the review. Please give the pywps-4.4 branch a try ... when there are no complains I can make the 4.4.0 release.

**Comment by tlvu** [comment]
@cehbrecht sorry I took a day off yesterday, was not able to look at this sooner.

So basically if we diff this branch with `master`, the only differences would be the jobqueue feature?  Everything else are backported?  We could merge this branch back into `master` and it should NOT make a difference?

**Comment by tlvu** [comment]
Here is how `master` would change today if `pywps-4.4` is merged into `master`: https://github.com/geopython/pywps/compare/20e1e254a3f7914e555fa89f363d1f6eb5f3895c...e10c99c2c8472287f9463bab49d85f5e7597cac2

Looks like all the changes from 4.2.4 forward are missing on `master`.  Should be a good thing to have on `master`.

**Comment by tlvu** [comment]
Here is the reverse diff (if `master` would merge into `pywps-4.4`) https://github.com/geopython/pywps/compare/e10c99c2c8472287f9463bab49d85f5e7597cac2...20e1e254a3f7914e555fa89f363d1f6eb5f3895c.

Looks like mostly the jobqueue but there are other changes that I am not familiar with.

**Comment by fmigneault** [comment]
I feel it would be a great idea to have first the resolution of python 2/3 with minor patches in 4.4. 
Then, another (maybe major?) version for jobqueue and co. features. 
This will greatly align with OGC-API / owslib eventual work (https://github.com/opengeospatial/joint-ogc-osgeo-asf-sprint-2021/issues/2) that will require the job items.

I have started pointing at `pywps-4.4` with this PR https://github.com/crim-ca/weaver/pull/214 
And so far it doesn't look like there are any major incompatibilities. 

I fear though that the alembic database migration and all other job queue elements would cause a lot of problems on my side, while being unnecessary in `weaver`'s case as jobs already have their own reporting/handling.

**Comment by cehbrecht** [comment]
> I fear though that the alembic database migration and all other job queue elements would cause a lot of problems on my side, while being unnecessary in `weaver`'s case as jobs already have their own reporting/handling.

@fmigneault In the next attempt for the job queue implementation I would prefer to life without alembic database migration. The database is essentially used for the queue and this is just temporary storage.

**Comment by cehbrecht** [comment]
> Here is how `master` would change today if `pywps-4.4` is merged into `master`: [20e1e25...e10c99c](https://github.com/geopython/pywps/compare/20e1e254a3f7914e555fa89f363d1f6eb5f3895c...e10c99c2c8472287f9463bab49d85f5e7597cac2)
> 
> Looks like all the changes from 4.2.4 forward are missing on `master`. Should be a good thing to have on `master`.

I could create a dev-branch from the current `master` and re-integrate the `pywps-4.4` branch. The `master` should then be the same as `pywps-4.4`.

**Comment by tlvu** [comment]
> I feel it would be a great idea to have first the resolution of python 2/3 with minor patches in 4.4.
> Then, another (maybe major?) version for jobqueue and co. features.

I am not suggesting merging `master` to `pywps-4.4` so no jobqueue on 4.4.  I am only suggesting merge `pywps-4.4` to `master` so `master` still contain all the fixes (return to the normal workflow before this divergence).

The 2 compare links I generated is simply for us to audit the changes between `master` and `pywps-4.4` branch.

**Comment by tlvu** [comment]
> > Here is how `master` would change today if `pywps-4.4` is merged into `master`: [20e1e25...e10c99c](https://github.com/geopython/pywps/compare/20e1e254a3f7914e555fa89f363d1f6eb5f3895c...e10c99c2c8472287f9463bab49d85f5e7597cac2)
> > Looks like all the changes from 4.2.4 forward are missing on `master`. Should be a good thing to have on `master`.
> 
> I could create a dev-branch from the current `master` and re-integrate the `pywps-4.4` branch. The `master` should then be the same as `pywps-4.4`.

I think we simply just need to return to the regular workflow.  Correct me if I am wrong but we used to merge all branches back to `master` so `master` always have all the cumulative changes?

So when we create new branches from `master`, we ensure the new branch is not missing anything from any previous published releases?

If what I understood is correctly, then after the merge of `pywps-4.4` to `master`, we can fix up the jobqueue on `master` to your liking @cehbrecht then cut a new branch from there for new release.  This assume on `master` the only thing new vs `pywps-4.4` is the jobqueue and nothing else.

**Comment by cehbrecht** [comment]
> > I feel it would be a great idea to have first the resolution of python 2/3 with minor patches in 4.4.
> > Then, another (maybe major?) version for jobqueue and co. features.
> 
> I am not suggesting merging `master` to `pywps-4.4` so no jobqueue on 4.4. I am only suggesting merge `pywps-4.4` to `master` so `master` still contain all the fixes (return to the normal workflow before this divergence).
> 
> The 2 compare links I generated is simply for us to audit the changes between `master` and `pywps-4.4` branch.

Right. The job-queue is not on `pywps-4.4` ... only in the history of master.

**Comment by cehbrecht** [comment]
> If what I understood is correct, then after the merge of `pywps-4.4` to `master`, we can fix up the jobqueue on `master` to your liking @cehbrecht then cut a new branch from there for new release.

When me make a new attempt on the job-queue I would do this on a dev branch. We can merge it to master when we are really sure we want to keep it :smile:

The next things that could happen on master (after re-integration of pywps-4.4):
* re-formatting using black?
* clean up tests
* work on rest api

**Comment by tlvu** [comment]
> When me make a new attempt on the job-queue I would do this on a dev branch.

Correct.  I did not literally meant any commits direct on `master`.  I mean PR to `master` instead of PR to `pywps-4.4`.

Sorry if my message was not clear.  Nothing should land on `master` without a PR.

**Pull request merged**